Feature/comments crud

kossme opened the pull request:
добавлена функция изменения и удаления своих комментариев, а также сортировка комментариев по новизне.

При удалении комментария, у которого остались ответы. он помечается как удаленный, но ответы остаются. 
В БД comments новый столбец deleted с boolean.

Функция "пожаловаться" интегрирована.

Исправлены ошибки

kossme committed:
добавлена функция изменения и удаления своих комментариев, а также сортировка комментариев по новизне.

Также перенес eventListener кнопки "купить" в productPage.html в саму кнопку: onclick="addToProductToCart(event)", т.к. eventListener выдавал ошибку при скрытой кнопке, когда пользователь незалогинен.

kossme committed:
Merge remote-tracking branch 'origin/dev' into feature/CommentsCRUD

# Conflicts:
#	src/main/resources/static/js/comments.js

kossme committed:
добавлена функция изменения и удаления своих комментариев, а также сортировка комментариев по новизне.

При удалении комментария, у которого остались ответы. он помечается как удаленный, но ответы остаются.
В БД comments новый столбец deleted с boolean.

Функция "пожаловаться" интегрирована.

kossme committed:
добавлена функция изменения и удаления своих комментариев, а также сортировка комментариев по новизне.

При удалении комментария, у которого остались ответы. он помечается как удаленный, но ответы остаются.
В БД comments новый столбец deleted с boolean.

Функция "пожаловаться" интегрирована.

Исправлены ошибки

kossme committed:
Merge branch 'dev' into feature/CommentsCRUD

KuGa-Do commented on basic_module/src/main/resources/static/js/comments.js:
```diff
@@ -1,171 +1,566 @@
+                document.getElementById("replyDisplayId" + commentData.id).before(replaceBlock);
+            }
+        }
     }
 
-    $(function () {
-        let productId = decodeURI(document.URL.substring(document.URL.lastIndexOf('/') + 1));
 
-        showComments();
-        $('#commentForm').on('submit', function (event) {
```
По идее Эдуард скажет удалить эту кучу пустых строк , но я не уверен ))) так что смотри сам

kossme replied:
Исправил

KuGa-Do reviewed (commented):
Функционал проверил , все как заявлено в сообщении пулреквеста, работает .
По код стайлу ,я откомментил  в каком месте ,  несколько пустых строк , скорей всего, я честно говоря не супер мастер код стайла, особенно по js  , но Эдуард скорей всего тебе об этом скажет , так что смотри сам исправлять тебе это или нет, еще там были места , где больше двух пустых строк , по идее это не очень хорошо .

kossme committed:
добавлена функция изменения и удаления своих комментариев, а также сортировка комментариев по новизне.

При удалении комментария, у которого остались ответы. он помечается как удаленный, но ответы остаются.
В БД comments новый столбец deleted с boolean.

Функция "пожаловаться" интегрирована.

Исправлены ошибки

Удалены лишние пробелы

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,51 @@
+            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+            String email = auth.getName();
+            User user = userService.findByEmail(email).orElse(new User());
+            System.out.println(user);
```
это точно убрать

kossme replied:
System.out.println(user);
Вот это? удалил

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,51 @@
+            if (commentService.findById(commentId).getCustomer().getId().equals(user.getId())) {
+                commentService.removeById(commentId);
+            } else {
+                throw new RuntimeException();
```
неверно использовать бросание exeption для построения логики приложения

kossme replied:
Переделал

kossme replied:
@DeleteMapping("/{commentId}")
    public ResponseEntity<?> deleteCommentById(@PathVariable Long commentId) {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        String email = auth.getName();
        if (userService.findByEmail(email).isPresent()) {
            User user = userService.findByEmail(email).get();
            if (commentService.findById(commentId).getCustomer().getId().equals(user.getId())) {
                commentService.removeById(commentId);
                return new ResponseEntity<>(HttpStatus.OK);
            }
        }
        return new ResponseEntity<>(HttpStatus.NOT_MODIFIED);
    }

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,51 @@
+        try {
+            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+            String email = auth.getName();
+            User user = userService.findByEmail(email).orElse(new User());
```
т.е. если мы не нашли пользователя, то просто создаем нового, а потом ниже пытаемся взять у него id?
Предлагаю пересмотреть логику этого метода

kossme replied:
В этом исполнении это неважно, т.к. кидалось бы исключение в любом случае при сравнении. Но согласен, что это коряво. Переделал

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,51 @@
+    @PutMapping
+    public ResponseEntity<?> updateComment(@RequestBody @Valid Comment comment) {
+        try {
+            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+            String email = auth.getName();
+            User user = userService.findByEmail(email).orElse(new User());
+            if (commentService.findById(comment.getId()).getCustomer().getId().equals(user.getId())) {
+                commentService.update(comment);
+            } else {
+                throw new RuntimeException();
+            }
+        } catch (Exception e) {
+            return new ResponseEntity<>(HttpStatus.NOT_MODIFIED);
+        }
```
смотреть замечания выше - переделать

tediik commented on basic_module/src/main/java/com/jm/online_store/model/Comment.java:
```diff
@@ -59,7 +61,16 @@
     @OneToMany(mappedBy = "comment", cascade = CascadeType.REMOVE)
     private List<ReportComment> reportComments;
 
+    @Column
+    @NonNull
+    private boolean deletedHaveKids;
+
```
да, но мы ведь можем определять если ли kids просто запрашивая поле reportComments и если Лист пустой, то и kids нету, разве нет? Зачем нам тогда еще одно поле в БД?

tediik commented on basic_module/src/main/java/com/jm/online_store/service/impl/CommentServiceImpl.java:
```diff
@@ -88,6 +88,30 @@ public Comment findById(Long commentId) {
     }
 
+
+    /**
+     * Delete ProductComment from database by comment Id
+     *
+     *
+     */
```
оформить JavaDoc как положено

tediik commented on basic_module/src/main/java/com/jm/online_store/service/impl/CommentServiceImpl.java:
```diff
@@ -88,6 +88,30 @@ public Comment findById(Long commentId) {
+        commentRepository.deleteById(commentId);
+    }
+
+    /**
+     * Обновляет комментарий. Из измененного коммента с фронта достается контент
+     * и присваивается этому же комментарию в БД
+     *
+     */
```
оформить JavaDoc как положено

tediik commented on basic_module/src/main/resources/static/js/productPage.js:
```diff
@@ -123,6 +123,7 @@ function addToFavourite() {
                 $("#heart").toggleClass("filled");
                 $("#favoriteLabel").empty().append(`<h5 class="font-weight-normal my-2">&ensp;В избранное</h5>`)
                 toastr.success("Товар успешно удалён из избранного");
+                toastr.success("Товар успешно удалён из избранного");
```
почему дублирование строк?

kossme replied:
Какой-то косяк, удалю

tediik commented on basic_module/src/main/resources/static/js/comments.js:
```diff
@@ -1,171 +1,524 @@
-                })
+        /*Если это удаленный комментарий с reply-ответами - вставляет заглушку*/
+        function isDeleted(commentData) {
+            if (commentData.deletedHaveKids) {
```
Причем чтобы не сломать логику в js, ты можешь добавить это булево поле в DTO, которое у тебя будет иметь значение в зависимости от того пустой ли список о котором я писал ранее

kossme committed:
Переделаны методы CommentRestController, исправлены ошибки в comment.js.

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,46 @@
+    public ResponseEntity<?> updateComment(@RequestBody @Valid Comment comment) {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        String email = auth.getName();
+        if (userService.findByEmail(email).isPresent()) {
+            User user = userService.findByEmail(email).get();
+            if (commentService.findById(comment.getId()).getCustomer().getId().equals(user.getId())) {
+                commentService.update(comment);
+                return new ResponseEntity<>(HttpStatus.OK);
```
не очень красивая работа с Optional, он намного интереснее чем ты думаешь, попробуй использовать метод ifPresent()

tediik commented on basic_module/src/main/java/com/jm/online_store/controller/rest/CommentRestController.java:
```diff
@@ -97,6 +108,46 @@
+    public ResponseEntity<?> deleteCommentById(@PathVariable Long commentId) {
+        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+        String email = auth.getName();
+        if (userService.findByEmail(email).isPresent()) {
+            User user = userService.findByEmail(email).get();
+            if (commentService.findById(commentId).getCustomer().getId().equals(user.getId())) {
```
попробуй использовать метод ifPresent()

kossme committed:
Изменены методы в CommentRestController: DELETE & PUT

kossme committed:
Merge branch 'dev' into feature/CommentsCRUD

kossme committed:
Изменены методы в CommentRestController: DELETE & PUT
Исправлены ошибки

kossme committed:
Merge branch 'dev' into feature/CommentsCRUD

kossme committed:
Merge branch 'dev' into feature/CommentsCRUD
